File: services/e2e/src/automations.test.ts
import { test, before, after, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import { pino } from 'pino';
import type { FastifyInstance } from 'fastify';
import { buildApp } from '@wolf/api/app';
import type { AppContext } from '@wolf/api/context';
import { InMemoryRateLimiter } from '@wolf/api/rate-limit';
import { AutomationExecutor, AutomationJob } from '@wolf/api/automation';
import { createRepositories, loadConfig, migrate, withTransaction, type CommandRecord } from '@wolf/server-core';
import { createTestDatabase, type TestDatabase } from '@wolf/server-core/testing';
import { createHmacSigner, hashPassword } from '@wolf/auth';
import { newId } from '@wolf/shared-types';

/**
 * Automations, end to end: saving one is authorizing it, and running one acts on that authority.
 *
 * The things that must hold, each a test: nothing critical can be automated; a high-risk automation
 * needs the same confirmation and password a person's command does; a run sends commands on the same
 * audited path; a cooldown and a scheduled minute are claimed once however many instances race; and
 * revoking the device an automation was saved from ends its authority.
 *
 * The agent is played by completing command rows directly, which is what the realtime service does
 * when an agent answers.
 */

const OWNER_EMAIL = 'automation-owner@example.com';
const OWNER_PASSWORD = 'a-long-owner-passphrase';

let db: TestDatabase;
let app: FastifyInstance;
let context: AppContext;
let userId: string;
let pcId: string;
let token: string;
let deviceId: string;
let clock: Date;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function login(name: string): Promise<string> {
  const response = await app.inject({
    method: 'POST',
    url: '/api/v1/auth/login',
    payload: { email: OWNER_EMAIL, password: OWNER_PASSWORD, device: { kind: 'web', name } },
  });
  assert.equal(response.statusCode, 200, response.payload);
  return String(JSON.parse(response.payload).accessToken);
}

before(async () => {
  db = await createTestDatabase();
  await migrate(db);

  const config = loadConfig({
    NODE_ENV: 'test',
    DATABASE_URL: 'postgres://test/wolf',
    WOLF_TOKEN_SECRET: 'e2e-secret-key-that-is-long-enough-to-pass',
    WOLF_ALLOWED_ORIGINS: 'http://localhost:3000',
    WOLF_COMMAND_TTL_SECONDS: '60',
  } as NodeJS.ProcessEnv);

  clock = new Date(Math.floor(Date.now() / 60_000) * 60_000);
  const repos = createRepositories(db);
  context = {
    config,
    db,
    repos,
    logger: pino({ level: 'silent' }),
    signer: createHmacSigner(config.tokens.secret),
    rateLimiter: new InMemoryRateLimiter(),
    now: () => clock,
  };

  const owner = await repos.users.createOwner({
    id: newId(),
    email: OWNER_EMAIL,
    passwordHash: await hashPassword(OWNER_PASSWORD, { cost: 2 ** 12, blockSize: 8, parallelization: 1, keyLength: 32 }),
    displayName: 'Owner',
  });
  userId = owner.id;

  app = await buildApp(context);
  await app.ready();

  token = await login('Automation browser');
  deviceId = (await repos.devices.listForUser(userId))[0]!.id;
});

beforeEach(async () => {
  clock = new Date(Math.floor(Date.now() / 60_000) * 60_000);
  await db.query('DELETE FROM automations');
  await db.query('DELETE FROM notifications');

  const pc = await context.repos.pcs.create({
    id: newId(),
    userId,
    name: `PC-${newId().slice(-4)}`,
    hostname: null,
    publicKey: 'a'.repeat(64),
    agentVersion: null,
  });
  pcId = pc.id;
  await db.query(`UPDATE pcs SET registration_state = 'active', status = 'online' WHERE id = $1`, [pcId]);
});

after(async () => {
  await app?.close();
  await db.end();
});

function call(method: 'GET' | 'POST' | 'PATCH' | 'DELETE', url: string, payload?: unknown, bearer = token) {
  return app.inject({
    method,
    url: `/api/v1${url}`,
    headers: { authorization: `Bearer ${bearer}` },
    ...(payload === undefined ? {} : { payload: payload as Record<string, unknown> }),
  });
}

const notify = { kind: 'notify', severity: 'info', message: 'Nightly check ran' };
const restartService = {
  kind: 'command',
  command: { type: 'service.control', payload: { name: 'Spooler', action: 'start', expectedDisplayName: 'Print Spooler' } },
};
const restartPc = {
  kind: 'command',
  command: { type: 'power.action', payload: { action: 'restart', delaySeconds: 60, force: false } },
};

function manual(actions: unknown[], overrides: Record<string, unknown> = {}) {
  return { name: 'By hand', trigger: { kind: 'manual' }, actions, targets: { mode: 'pcs', pcIds: [pcId] }, ...overrides };
}

async function create(automation: unknown, confirmedRiskLevel?: string) {
  const response = await call('POST', '/automations', { automation, confirmedRiskLevel });
  assert.equal(response.statusCode, 201, response.payload);
  return JSON.parse(response.payload).automation as { id: string; enabled: boolean; authorizedRiskLevel: string };
}

async function runs(automationId: string) {
  return context.repos.automations.listRuns(automationId, userId);
}

async function waitForRun(automationId: string, done: (run: Awaited<ReturnType<typeof runs>>[number]) => boolean) {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const found = (await runs(automationId)).find(done);
    if (found) return found;
    await sleep(50);
  }
  throw new Error('the run never reached the expected state');
}

/** Play the agent: complete the next command of a type as soon as it is dispatched. */
async function answerNextCommand(type: string, status: 'completed' | 'failed' = 'completed'): Promise<CommandRecord> {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const pending = (await context.repos.commands.listForPc(pcId, 20)).find(
      (command) => command.type === type && command.status === 'pending',
    );
    if (pending) {
      await context.repos.commands.complete({
        id: pending.id,
        status,
        startedAt: clock,
        completedAt: clock,
        errorCode: status === 'failed' ? 'access-denied' : null,
        errorMessage: null,
        errorIsLimitation: false,
        result: status === 'completed' ? { ok: true } : null,
      });
      return pending;
    }
    await sleep(25);
  }
  throw new Error(`no ${type} command was dispatched`);
}

/* ------------------------------------------------------------------------- */
/* Saving is authorizing                                                      */
/* ------------------------------------------------------------------------- */

test('automations require a signed-in caller', async () => {
  for (const [method, url] of [
    ['GET', '/api/v1/automations'],
    ['POST', '/api/v1/automations'],
  ] as const) {
    assert.equal((await app.inject({ method, url })).statusCode, 401);
  }
});

test('what cannot be automated is refused at save time', async () => {
  const refusals = [
    // Names a live PID: a PID saved today is a different process tomorrow.
    manual([{ kind: 'command', command: { type: 'process.terminate', payload: { pid: 1234, expectedName: 'notepad.exe' } } }]),
    // Forced restart is critical.
    manual([{ kind: 'command', command: { type: 'power.action', payload: { action: 'restart', force: true } } }]),
    // "The alert's PC" without an alert.
    manual([notify], { targets: { mode: 'alert-pc' } }),
    { ...manual([notify]), trigger: { kind: 'schedule', time: '03:00', days: ['mon'], timeZone: 'Mars/Olympus_Mons' } },
    manual([]),
  ];

  for (const automation of refusals) {
    const response = await call('POST', '/automations', { automation, confirmedRiskLevel: 'critical' });
    assert.equal(response.statusCode, 400, response.payload);
  }

  const strangerPc = await call('POST', '/automations', { automation: manual([notify], { targets: { mode: 'pcs', pcIds: [newId()] } }) });
  assert.equal(strangerPc.statusCode, 404);
});

test('a notification-only automation saves without confirmation and is audited', async () => {
  const automation = await create(manual([notify]));
  assert.equal(automation.authorizedRiskLevel, 'low');

  const { rows } = await db.query<{ after_value: { actions: string[] } }>(
    `SELECT after_value FROM audit_logs WHERE action = 'automation.create' AND target->>'automationId' = $1`,
    [automation.id],
  );
  assert.deepEqual(rows[0]?.after_value.actions, ['notify']);
  // The owner's message is not copied into the audit trail.
  assert.ok(!JSON.stringify(rows[0]).includes('Nightly check ran'));
});

test('a high-risk automation needs the matching confirmation and a fresh password', async () => {
  const automation = manual([notify, restartPc]);

  const unconfirmed = await call('POST', '/automations', { automation });
  assert.equal(unconfirmed.statusCode, 428);
  assert.equal(JSON.parse(unconfirmed.payload).error.code, 'command.confirmation_required');
  assert.equal(JSON.parse(unconfirmed.payload).error.context.riskLevel, 'high');

  const milder = await call('POST', '/automations', { automation, confirmedRiskLevel: 'medium' });
  assert.equal(milder.statusCode, 428);

  // Six minutes after signing in, the password is too old for a high-risk decision.
  clock = new Date(clock.getTime() + 6 * 60_000);
  const stale = await call('POST', '/automations', { automation, confirmedRiskLevel: 'high' });
  assert.equal(stale.statusCode, 428);
  assert.equal(JSON.parse(stale.payload).error.code, 'command.reauth_required');

  const fresh = await login('Fresh password');
  const saved = await call('POST', '/automations', { automation, confirmedRiskLevel: 'high' }, fresh);
  assert.equal(saved.statusCode, 201, saved.payload);
  assert.equal(JSON.parse(saved.payload).automation.authorizedRiskLevel, 'high');
});

test('renaming or turning off needs nothing; widening re-authorizes', async () => {
  const automation = await create(manual([notify]));

  assert.equal((await call('PATCH', `/automations/${automation.id}`, { automation: { name: 'Renamed' } })).statusCode, 200);
  assert.equal((await call('PATCH', `/automations/${automation.id}`, { automation: { enabled: false } })).statusCode, 200);

  // Adding a service action is a new decision.
  const widened = await call('PATCH', `/automations/${automation.id}`, { automation: { actions: [notify, restartService], enabled: true } });
  assert.equal(widened.statusCode, 428);

  const confirmed = await call('PATCH', `/automations/${automation.id}`, {
    automation: { actions: [notify, restartService], enabled: true },
    confirmedRiskLevel: 'medium',
  });
  assert.equal(confirmed.statusCode, 200, confirmed.payload);
  assert.equal(JSON.parse(confirmed.payload).automation.authorizedRiskLevel, 'medium');
});

/* ------------------------------------------------------------------------- */
/* Running                                                                    */
/* ------------------------------------------------------------------------- */

test('a run notifies, sends its command on the audited path, and waits for the result', async () => {
  const automation = await create(manual([notify, restartService]), 'medium');

  const accepted = await call('POST', `/automations/${automation.id}/run`);
  assert.equal(accepted.statusCode, 202);

  const command = await answerNextCommand('service.control');
  const run = await waitForRun(automation.id, (candidate) => candidate.status === 'completed');

  assert.equal(run.triggerKind, 'manual');
  assert.deepEqual(run.steps.map((step) => [step.kind, step.status]), [['notify', 'completed'], ['command', 'completed']]);
  assert.equal(run.steps[1]?.commandId, command.id);

  // No session, and the confirmation is dated when it was given: at save time.
  assert.equal(command.sessionId, null);
  assert.equal(command.deviceId, deviceId);
  assert.equal(command.riskLevel, 'medium');
  assert.ok(command.authorization.confirmedAt);
  assert.deepEqual(command.authorization.grantedCapabilities, ['services']);

  const { rows } = await db.query<{ action: string; outcome: string }>(
    `SELECT action, outcome FROM audit_logs WHERE request_id = $1 ORDER BY occurred_at`,
    [run.id],
  );
  assert.deepEqual(rows.map((row) => `${row.action}:${row.outcome}`).sort(), ['automation.run:success', 'service.control:pending']);

  const inbox = await context.repos.alerts.listNotifications(userId, { unreadOnly: false, limit: 10 });
  assert.equal(inbox.length, 1);
  assert.equal(inbox[0]?.kind, 'automation');
  assert.equal(inbox[0]?.automationId, automation.id);
});

test('a failed command stops the run, skips the rest and tells the owner', async () => {
  const automation = await create(manual([restartService, notify]), 'medium');

  await call('POST', `/automations/${automation.id}/run`);
  await answerNextCommand('service.control', 'failed');
  const run = await waitForRun(automation.id, (candidate) => candidate.status === 'failed');

  assert.deepEqual(run.steps.map((step) => step.status), ['failed', 'skipped']);
  assert.equal(run.reason, 'access-denied');

  const inbox = await context.repos.alerts.listNotifications(userId, { unreadOnly: false, limit: 10 });
  assert.equal(inbox.length, 1, 'the failure, and not the skipped notify action');
  assert.match(inbox[0]!.title, /did not finish/);
});

test('an offline PC is refused before anything is queued, and that is not news', async () => {
  await db.query(`UPDATE pcs SET status = 'offline' WHERE id = $1`, [pcId]);
  const automation = await create(manual([restartService]), 'medium');

  await call('POST', `/automations/${automation.id}/run`);
  const run = await waitForRun(automation.id, (candidate) => candidate.status === 'failed');

  assert.equal(run.reason, 'pc-offline');
  assert.equal((await context.repos.commands.listForPc(pcId)).length, 0, 'never queued to run late');
  assert.equal((await context.repos.alerts.listNotifications(userId, { unreadOnly: false, limit: 10 })).length, 0);

  const { rows } = await db.query(`SELECT 1 FROM audit_logs WHERE outcome = 'denied' AND error_code = 'pc-offline' AND request_id = $1`, [run.id]);
  assert.equal(rows.length, 1);
});

test('a live session holding power control is never overridden', async () => {
  const automation = await create(manual([restartPc]), 'high');

  // Somebody is connected and holding the PC's power control.
  const sessionId = newId();
  await context.repos.sessions.create({
    id: sessionId,
    userId,
    deviceId,
    pcId,
    mode: 'control',
    capabilities: ['power'],
    route: 'relay',
    expiresAt: new Date(Date.now() + 600_000),
  });
  await context.repos.sessions.acquireResource({ pcId, resource: 'power', sessionId, expiresAt: new Date(Date.now() + 600_000) });

  await call('POST', `/automations/${automation.id}/run`);
  const run = await waitForRun(automation.id, (candidate) => candidate.status === 'failed');
  assert.equal(run.reason, 'resource-held');
});

/* ------------------------------------------------------------------------- */
/* Cooldowns and limits                                                       */
/* ------------------------------------------------------------------------- */

test('two instances reaching the same automation start one run; the other is cooling down', async () => {
  const saved = await create(manual([notify], { cooldownMinutes: 30 }));
  const automation = (await context.repos.automations.findById(saved.id, userId))!;

  const first = new AutomationExecutor(context);
  const second = new AutomationExecutor(context);
  first.start(automation, [pcId], 'schedule');
  second.start(automation, [pcId], 'schedule');
  await Promise.all([first.drain(), second.drain()]);

  const statuses = (await runs(saved.id)).map((run) => `${run.status}:${run.reason ?? ''}`).sort();
  assert.deepEqual(statuses, ['completed:', 'skipped:cooldown']);
});

test('the daily limit bounds even runs started by hand', async () => {
  const saved = await create(manual([notify], { maxRunsPerDay: 1 }));

  await call('POST', `/automations/${saved.id}/run`);
  await waitForRun(saved.id, (run) => run.status === 'completed');
  await call('POST', `/automations/${saved.id}/run`);
  const limited = await waitForRun(saved.id, (run) => run.status === 'skipped');

  assert.equal(limited.reason, 'daily-limit');
});

test('conditions are checked before the cooldown is used', async () => {
  const saved = await create(manual([notify], { conditions: [{ kind: 'metric', metric: 'cpu.usage', comparison: 'below', threshold: 20 }] }));
  const automation = (await context.repos.automations.findById(saved.id, userId))!;

  const executor = new AutomationExecutor(context);
  executor.start(automation, [pcId], 'schedule');
  await executor.drain();

  const [run] = await runs(saved.id);
  assert.equal(run?.status, 'skipped');
  assert.match(run!.reason!, /^condition-not-met: No recent telemetry/);
  const { rows } = await db.query('SELECT 1 FROM automation_pc_state WHERE automation_id = $1', [saved.id]);
  assert.equal(rows.length, 0, 'a run that never acted did not start the cooldown');
});

/* ------------------------------------------------------------------------- */
/* Authority ends                                                             */
/* ------------------------------------------------------------------------- */

test('an automation whose authorizing device was revoked turns itself off and says so', async () => {
  const saved = await create(manual([notify]));
  const automation = (await context.repos.automations.findById(saved.id, userId))!;

  // Revoked underneath the automation, as if by a path that did not disable it.
  await db.query(`UPDATE user_devices SET status = 'revoked', revoked_at = now() WHERE id = $1`, [deviceId]);
  try {
    const executor = new AutomationExecutor(context);
    executor.start(automation, [pcId], 'schedule');
    await executor.drain();

    const [run] = await runs(saved.id);
    assert.equal(run?.status, 'skipped');
    assert.match(run!.reason!, /^authority-revoked/);
    assert.equal((await context.repos.automations.findById(saved.id, userId))?.enabled, false);

    const inbox = await context.repos.alerts.listNotifications(userId, { unreadOnly: false, limit: 10 });
    assert.match(inbox[0]!.title, /was turned off/);
  } finally {
    await db.query(`UPDATE user_devices SET status = 'active', revoked_at = NULL WHERE id = $1`, [deviceId]);
  }
});

test('revoking a device through the API turns off its automations in the same transaction', async () => {
  const other = await login('Second device');
  const devices = await context.repos.devices.listForUser(userId);
  const target = devices.find((device) => device.name === 'Second device' && device.status === 'active')!;

  const saved = await create(manual([notify]));
  // Authorized from the second device.
  await db.query('UPDATE automations SET authorized_device_id = $2 WHERE id = $1', [saved.id, target.id]);

  const revoked = await call('DELETE', `/devices/${target.id}`);
  assert.equal(revoked.statusCode, 204, revoked.payload);
  void other;

  assert.equal((await context.repos.automations.findById(saved.id, userId))?.enabled, false);
});

/* ------------------------------------------------------------------------- */
/* Triggers                                                                   */
/* ------------------------------------------------------------------------- */

test('a scheduled minute runs once, however many instances reach it', async () => {
  const time = clock.toISOString().slice(11, 16);
  const saved = await create({
    ...manual([notify]),
    trigger: { kind: 'schedule', time, days: ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'], timeZone: 'UTC' },
  });

  const jobs = [new AutomationJob(context), new AutomationJob(context)];
  await Promise.all(jobs.map((job) => job.runOnce()));
  await Promise.all(jobs.map((job) => job.executor.drain()));

  // A minute later, still within the grace window: the slot is already claimed.
  clock = new Date(clock.getTime() + 60_000);
  await jobs[0]!.runOnce();
  await jobs[0]!.executor.drain();

  assert.deepEqual((await runs(saved.id)).map((run) => run.status), ['completed']);
});

test('a schedule not due now does nothing', async () => {
  const other = new Date(clock.getTime() + 3 * 3_600_000).toISOString().slice(11, 16);
  const saved = await create({ ...manual([notify]), trigger: { kind: 'schedule', time: other, days: ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'], timeZone: 'UTC' } });

  const job = new AutomationJob(context);
  await job.runOnce();
  await job.executor.drain();

  assert.equal((await runs(saved.id)).length, 0);
});

test('an alert firing runs the automation on the PC it fired for; a stale one does not', async () => {
  const rule = await context.repos.alerts.createRule(userId, {
    pcId: null,
    name: 'Offline',
    condition: 'pc-offline',
    metric: null,
    seriesKey: null,
    threshold: null,
    forMinutes: 10,
    severity: 'warning',
    cooldownMinutes: 60,
    enabled: true,
  });

  const saved = await create({
    name: 'On alert',
    trigger: { kind: 'alert', ruleId: rule.id, on: 'fired' },
    actions: [notify],
    targets: { mode: 'alert-pc' },
  });

  const record = (occurredAt: Date) =>
    withTransaction(db, (client) =>
      context.repos.automations.recordAlertEvent(client, { userId, ruleId: rule.id, pcId, kind: 'alert-fired', occurredAt }),
    );

  assert.equal(await record(clock), true);
  // A resolution is not what this automation asks about.
  await withTransaction(db, (client) =>
    context.repos.automations.recordAlertEvent(client, { userId, ruleId: rule.id, pcId, kind: 'alert-resolved', occurredAt: clock }),
  );

  const job = new AutomationJob(context);
  await job.runOnce();
  await job.executor.drain();

  let history = await runs(saved.id);
  assert.equal(history.length, 1);
  assert.equal(history[0]?.pcId, pcId);
  assert.equal(history[0]?.triggerKind, 'alert');

  await record(new Date(clock.getTime() - 20 * 60_000));
  await job.runOnce();
  await job.executor.drain();

  history = await runs(saved.id);
  assert.ok(history.some((run) => run.status === 'skipped' && run.reason?.startsWith('stale-trigger')));
});

File: apps/android/app/src/test/java/app/amizhthan/wolf/api/AutomationsTest.kt
package app.amizhthan.wolf.api

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

/** The automations and alert rules the phone builds, held to `packages/protocol` shape and bounds. */
class AutomationsTest {
    private val pc = "01J9ZQK7T0000000000000000P"

    private fun refused(block: () -> Unit): String = try {
        block()
        fail("expected a refusal")
        ""
    } catch (error: IllegalArgumentException) {
        error.message.orEmpty()
    }

    @Test
    fun an_automation_is_the_protocol_shape_field_for_field() {
        val definition = Automations.definition(
            name = "  Nightly restart ",
            trigger = Automations.schedule("03:00", Automations.DAYS, "UTC"),
            conditions = listOf(Automations.noActiveSession()),
            actions = listOf(Automations.power("restart", 60)),
            targets = Automations.onPcs(listOf(pc)),
        )

        assertEquals(
            """{"name":"Nightly restart","enabled":true,"trigger":{"kind":"schedule","time":"03:00","days":["mon","tue","wed","thu","fri","sat","sun"],"timeZone":"UTC"},""" +
                """"conditions":[{"kind":"no-active-session"}],"actions":[{"kind":"command","command":{"type":"power.action","payload":{"action":"restart","delaySeconds":60,"force":false}}}],""" +
                """"targets":{"mode":"pcs","pcIds":["$pc"]},"cooldownMinutes":60,"maxRunsPerDay":4}""",
            definition.toString(),
        )
    }

    @Test
    fun days_are_sent_in_week_order_and_times_are_24_hour() {
        assertEquals(
            """{"kind":"schedule","time":"23:59","days":["mon","sun"],"timeZone":"Europe/London"}""",
            Automations.schedule("23:59", listOf("sun", "mon"), "Europe/London").toString(),
        )
        listOf("3:00", "24:00", "12:60", "noon", "").forEach { time -> refused { Automations.schedule(time, Automations.DAYS, "UTC") } }
        refused { Automations.schedule("03:00", emptyList(), "UTC") }
        refused { Automations.schedule("03:00", listOf("someday"), "UTC") }
        refused { Automations.timeWindow("22:00", "6:00", Automations.DAYS, "UTC") }
    }

    @Test
    fun an_alert_trigger_for_any_rule_says_so_with_null() {
        assertEquals("""{"kind":"alert","ruleId":null,"on":"fired"}""", Automations.onAlert(null, "fired").toString())
        refused { Automations.onAlert(null, "changed") }
    }

    @Test
    fun the_pc_the_alert_fired_for_needs_an_alert_trigger() {
        val message = refused {
            Automations.definition("x", Automations.manual(), emptyList(), listOf(Automations.notify("hi", "info")), Automations.alertPc())
        }
        assertTrue(message.contains("alert trigger"))

        Automations.definition("x", Automations.onAlert(null, "fired"), emptyList(), listOf(Automations.notify("hi", "info")), Automations.alertPc())
    }

    @Test
    fun an_automation_is_bounded_the_way_the_server_bounds_it() {
        val notify = Automations.notify("hi", "info")
        val targets = Automations.onPcs(listOf(pc))
        fun build(name: String = "x", actions: Int = 1, conditions: Int = 0, cooldown: Int = 60, runs: Int = 4) = Automations.definition(
            name, Automations.manual(), List(conditions) { Automations.noActiveSession() }, List(actions) { notify }, targets, cooldown, runs,
        )

        build(actions = 5, conditions = 5, cooldown = 10_080, runs = 96)
        refused { build(actions = 0) }
        refused { build(actions = 6) }
        refused { build(conditions = 6) }
        refused { build(cooldown = 0) }
        refused { build(runs = 97) }
        refused { build(name = "   ") }
        refused { build(name = "n".repeat(121)) }
        refused { Automations.onPcs(emptyList()) }
        refused { Automations.onPcs(listOf(pc, pc)) }
        refused { Automations.onPcs(List(21) { "01J9ZQK7T00000000000000${it.toString().padStart(3, '0')}" }) }
    }

    @Test
    fun power_actions_are_never_forced_and_wait_at_most_a_day() {
        val command = Automations.power("shutdown", 86_400)["command"]!!.jsonObject
        assertEquals("false", command["payload"]!!.jsonObject["force"].toString())
        assertEquals("86400", command["payload"]!!.jsonObject["delaySeconds"].toString())

        refused { Automations.power("shutdown", -1) }
        refused { Automations.power("shutdown", 86_401) }
        refused { Automations.power("format", 0) }
        // The PC screen's power buttons still send no delay.
        assertEquals("0", Commands.power("lock")["payload"]!!.jsonObject["delaySeconds"].toString())
    }

    @Test
    fun a_notification_is_trimmed_and_bounded() {
        assertEquals("""{"kind":"notify","severity":"warning","message":"Disk nearly full"}""", Automations.notify("  Disk nearly full ", "warning").toString())
        refused { Automations.notify("   ", "info") }
        refused { Automations.notify("m".repeat(201), "info") }
        refused { Automations.notify("hi", "urgent") }
        refused { Automations.cpuBelow(0.0) }
        refused { Automations.cpuBelow(Double.NaN) }
    }

    @Test
    fun automations_are_described_in_the_dashboard_words() {
        val rules = mapOf("01J9ZQK7T0000000000000000R" to "Disk nearly full")
        assertEquals("At 03:00 on every day (UTC)", Automations.describeTrigger(Automations.schedule("03:00", Automations.DAYS, "UTC")) { rules[it] })
        assertEquals("At 22:30 on sat, sun (UTC)", Automations.describeTrigger(Automations.schedule("22:30", listOf("sat", "sun"), "UTC")) { rules[it] })
        assertEquals("When Disk nearly full fires", Automations.describeTrigger(Automations.onAlert("01J9ZQK7T0000000000000000R", "fired")) { rules[it] })
        assertEquals("When a deleted rule resolves", Automations.describeTrigger(Automations.onAlert("01J9ZQK7T0000000000000000X", "resolved")) { rules[it] })
        assertEquals("When any alert rule fires", Automations.describeTrigger(Automations.onAlert(null, "fired")) { rules[it] })
        assertEquals("Only when run by hand", Automations.describeTrigger(Automations.manual()) { null })

        assertEquals("restart the PC after 60 s", Automations.describeAction(Automations.power("restart", 60)))
        assertEquals("lock the PC", Automations.describeAction(Automations.power("lock", 0)))
        assertEquals("notify \"Hello\"", Automations.describeAction(Automations.notify("Hello", "info")))
        val service = buildJsonObject {
            put("kind", "command")
            putJsonObject("command") {
                put("type", "service.control")
                putJsonObject("payload") {
                    put("name", "Spooler")
                    put("action", "restart")
                }
            }
        }
        assertEquals("restart service Spooler", Automations.describeAction(service))

        assertEquals("DESKTOP, LAPTOP", Automations.describeTargets(Automations.onPcs(listOf("a".repeat(26), "b".repeat(26)))) { if (it.startsWith("a")) "DESKTOP" else "LAPTOP" })
        assertEquals("the PC the alert fired for", Automations.describeTargets(Automations.alertPc()) { it })
        assertEquals("CPU usage is below 20%", Automations.describeCondition(Automations.cpuBelow(20.0)))
        assertEquals("it is between 22:00 and 06:00", Automations.describeCondition(Automations.timeWindow("22:00", "06:00", Automations.DAYS, "UTC")))
    }

    @Test
    fun what_this_app_does_not_model_is_shown_by_name_not_dropped() {
        val listed = WolfJson.decodeFromString(
            AutomationList.serializer(),
            """{"automations":[{"id":"01J9ZQK7T0000000000000000A","name":"Future","enabled":true,"trigger":{"kind":"webhook","path":"/x"},"conditions":[{"kind":"on-battery"}],"actions":[{"kind":"script"}],"targets":{"mode":"tagged","tag":"lab"},"cooldownMinutes":60,"maxRunsPerDay":4,"authorizedRiskLevel":"medium","authorizedAt":"2026-09-15T00:00:00Z","createdAt":"2026-09-15T00:00:00Z","updatedAt":"2026-09-15T00:00:00Z"}],"limit":50}""",
        ).automations.single()

        assertTrue(Automations.describeTrigger(listed.trigger) { null }.contains("webhook"))
        assertEquals("on-battery", Automations.describeCondition(listed.conditions.single()))
        assertEquals("script", Automations.describeAction(listed.actions.single()))
        assertEquals("tagged", Automations.describeTargets(listed.targets) { it })
        assertFalse(Automations.runsByHand(listed.targets))
        assertNull(listed.lastRunAt)
    }

    @Test
    fun run_reasons_are_worded_and_unknown_ones_passed_through() {
        assertEquals("A condition was not met — cpu.usage is 64", Automations.describeReason("condition-not-met: cpu.usage is 64"))
        assertEquals("The PC was offline", Automations.describeReason("pc-offline"))
        assertEquals("brand-new-reason", Automations.describeReason("brand-new-reason"))
        assertEquals("service.control: failed (service.not_found)", Automations.describeStep(AutomationStepView(0, "command", "failed", null, "service.control", "service.not_found")))
    }

    @Test
    fun an_offline_rule_carries_no_metric_and_a_metric_rule_needs_one() {
        val offline = AlertRules.rule("PC gone", null, "pc-offline", "cpu.usage", 90.0, "C:", 10, "critical", 60)
        assertNull(offline.metric)
        assertNull(offline.threshold)
        // Nulls are left out, so the server applies its own defaults for them.
        assertEquals(
            """{"name":"PC gone","condition":"pc-offline","forMinutes":10,"severity":"critical","cooldownMinutes":60,"enabled":true}""",
            WolfJson.encodeToString(AlertRuleInput.serializer(), offline),
        )

        refused { AlertRules.rule("Hot", pc, "metric-above", "cpu.temperature", null, null, 10, "warning", 60) }
        refused { AlertRules.rule("Hot", pc, "metric-above", "cpu.fanSpeed", 90.0, null, 10, "warning", 60) }
        refused { AlertRules.rule("Hot", pc, "metric-sideways", "cpu.usage", 90.0, null, 10, "warning", 60) }
        refused { AlertRules.rule("Hot", pc, "metric-above", "cpu.usage", 90.0, null, 0, "warning", 60) }
    }

    @Test
    fun a_device_is_kept_only_for_metrics_that_have_devices() {
        assertNull(AlertRules.rule("CPU", pc, "metric-above", "cpu.usage", 90.0, "C:", 10, "warning", 60).seriesKey)
        assertEquals("C:", AlertRules.rule("Disk", pc, "metric-above", "disk.usedPercent", 90.0, "  C: ", 10, "warning", 60).seriesKey)
        assertNull(AlertRules.rule("Disk", pc, "metric-above", "disk.usedPercent", 90.0, "   ", 10, "warning", 60).seriesKey)
    }

    @Test
    fun the_quiet_period_keeps_the_server_floor() {
        refused { AlertRules.rule("Flappy", null, "pc-offline", null, null, null, 10, "info", 4) }
        AlertRules.rule("Flappy", null, "pc-offline", null, null, null, 10, "info", 5)
    }

    @Test
    fun rules_and_notifications_are_described() {
        val disk = AlertRuleView("01J9ZQK7T0000000000000000R", pc, "Disk", "metric-above", "disk.usedPercent", "C:", 90.0, 30, "warning", 120, true, "t", "t")
        assertEquals("Disk used (C:) on DESKTOP above 90% for 30 min · quiet 120 min after notifying", AlertRules.describe(disk, "DESKTOP"))
        assertEquals(
            "every PC offline for 10 min · quiet 60 min after notifying",
            AlertRules.describe(disk.copy(condition = "pc-offline", metric = null, threshold = null, forMinutes = 10, cooldownMinutes = 60), "every PC"),
        )
        assertEquals(
            "Battery charge on LAPTOP below 12.5% for 5 min · quiet 60 min after notifying",
            AlertRules.describe(disk.copy(condition = "metric-below", metric = "battery.charge", seriesKey = null, threshold = 12.5, forMinutes = 5, cooldownMinutes = 60), "LAPTOP"),
        )

        val fired = NotificationView("n", kind = "fired", severity = "critical", title = "t", occurredAt = "t")
        assertEquals("Critical", AlertRules.label(fired))
        assertEquals("Resolved", AlertRules.label(fired.copy(kind = "resolved")))
        assertEquals("Automation · info", AlertRules.label(fired.copy(kind = "automation", severity = "info")))
    }
}
